src/ucc/api/ucc.h

```diff
@@ -1574,5 +1588,226 @@ static inline ucc_status_t ucc_collective_test(ucc_coll_req_h request)
+} ucc_ev_t;
+
+/**
+ * @ingroup UCC_EVENT_DT
+ *
+ */
+typedef struct ucc_ee_params {
+     ucc_ee_type_t  ee_type;
+     void           *ee_context;
+     size_t         ee_context_size;
+} ucc_ee_params_t;
+
+/**
+ * @ingroup UCC_EVENT
+ *
+ * @brief The routine creates the execution context for collective operations.
+ *
+ * @param [in] team     team handle
+ * @param [in] params   user provided params to customize the execution engine
+ * @param [out] ee      execution engine handle
+ *
+ * @parblock
+ *
+ * @b Description
+ *
+ * @ref ucc_ee_create creates the execution engine. It enables event-driven
+ * collective execution. @ref ucc_ee_params_t allows the execution engine to be
+ * configured to abstract either GPU and CPU threads. The execution engine is
+ * created and coupled with the team. There can be many execution engines
+ * coupled to the team. However, attaching the same execution engine to multiple
+ * teams is not allowed. The execution engine is created after the team is
+ * created and destroyed before the team is destroyed. It is the user's
+ * responsibility to destroy the execution engines before the team. If the team
+ * is destroyed before the execution engine is destroyed, the result is
+ * undefined.
+ *
+ * @endparblock
+ *
+ * @return Error code as defined by ucc_status_t
+ */
+ ucc_status_t ucc_ee_create(ucc_team_h team, const ucc_ee_params_t *params, ucc_ee_h *ee);
+
+ /**
+ * @ingroup UCC_EVENT
+ *
+ * @brief The routine destroys the execution context created for collective operations.
+ *
+ * @param [in] ee   Execution engine handle
+ *
+ * @parblock
+ *
+ * @b Description
+ *
+ * @ref ucc_ee_destroy releases the resources attached with the
+ * execution engine and destroys the execution engine. All events and triggered
+ * operations related to this ee are invalid after the destroy operation. To
+ * avoid race between the creation and destroying the execution engine, for a
+ * given ee, the @ref ucc_ee_create and @ref ucc_ee_destroy must be invoked from
+ * the same thread.
+ *
+ * @endparblock
+ *
+ * @return Error code as defined by ucc_status_t
+ */
+ ucc_status_t ucc_ee_destroy(ucc_ee_h ee);
+
+/**
+ * @ingroup UCC_EVENT
+ *
+ * @brief The routine gets the event from the event queue.
+ *
+ * @param [in]  ee        execution engine handle
+ * @param [out] ev        Event structure fetched from the event queue
+ *
+ * @parblock
+ *
+ * @b Description
+ *
+ * @ref ucc_ee_get_event fetches the events from the execution engine. If there
+ * are no events posted on the ee, it returns immediately without waiting for
+ * events. All events must be acknowledged using the @ref ucc_ee_ack_event
+ * interface. The event acknowledged is destroyed by the library. An event
+ * fetched with @ref ucc_ee_get_event but not acknowledged might consume
+ * resources in the library.
+ *
+ * @endparblock
+ *
+ * @return Error code as defined by ucc_status_t
+ */
+ucc_status_t ucc_ee_get_event(ucc_ee_h ee, ucc_ev_t *ev);
```

Comment:
is it correct that the only difference between ee_wait and ee_get_event is that ee_get_event is non blocking?

Reply:
Correct.